File: src/csharp/Foundgine.Core/Semantic/Resolution/IAdvancedCandidateSource.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Resolution;

/// <summary>Optional candidate-source extensions for richer semantic resolution.</summary>
public interface IAdvancedCandidateSource : ICandidateSource
{
    IReadOnlyList<IdentityCandidate> FindByCompositeIdentity(
        EntityId entityType,
        IReadOnlyDictionary<string, string> identityValues);

    IReadOnlyList<IdentityCandidate> FindByTemporalIdentity(
        EntityId entityType,
        string identityValue,
        DateTimeOffset asOf);
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/resolution/IAdvancedCandidateSource.java
package com.foundgine.core.semantic.resolution;

import com.foundgine.core.abstractions.*;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;

public interface IAdvancedCandidateSource extends ICandidateSource {
	List<IdentityCandidate> findByCompositeIdentity(EntityId entityType, Map<String, String> identityValues);

	List<IdentityCandidate> findByTemporalIdentity(EntityId entityType, String identityValue, OffsetDateTime asOf);
}
